state/upgradeseries.go

```diff
@@ -0,0 +1,56 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"github.com/juju/collections/set"
+	"github.com/juju/errors"
+	"gopkg.in/juju/names.v2"
+)
+
+// UpgradeSeriesLockApplicationIntersect returns application names represented
+// by the input machine's units that also have units on other machines that are
+// locked for upgrade.
+func (st *State) UpgradeSeriesLockApplicationIntersect(machineId string) ([]string, error) {
+	locks, err := st.getAllUpgradeSeriesLocks()
+	if err != nil {
+		return nil, errors.Annotatef(err,
+			"retrieving intersecting upgrade series applications for machine %v", machineId)
+	}
+
+	machineApps := set.NewStrings()
+	otherApps := set.NewStrings()
+
+	// Use the lock's unit record map to accrue sets for the input machine,
+	// and for all others that are currently locked.
+	for _, lock := range locks {
+		apps := set.NewStrings()
+		for unit := range lock.UnitStatuses {
+			app, err := names.UnitApplication(unit)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			apps.Add(app)
+		}
+
+		if machineId == lock.Id {
+			machineApps = machineApps.Union(apps)
```

Comment:
Some thoughts:

1) Can we have >1 lock for a single machine? If not, then we just have machineApps = apps, don't we?
2) My understanding is that this query is about "find if any of these applications should stay locked because they are being locked by another machine with a unit of the same application".
Which I think this satisfies. (get all apps that are currently locked on any machine that isn't this one, and compare it to the apps that this machine wants to lock.)
3) This *doesn't* distinguish completed locks, though maybe those won't be found in the getAllUpgradeSeriesLocks(). Though do we have any transactional guarantees? What happens if 2 machines are completed simultaneously? Will machine 0 see that machine 1 wants to still hold the lock, and machine 1 will see that machine 0 wants to hold the lock, and then neither of them unlock, though both of them proceed to the completed phase?

Reply:
1. Yes, correct. I should also add a trapdoor for when there is only one lock.
2. Yes.
3. It _should_ be safe given that the _upgradeseries_ commands are synchronous, and lock removal is part of the completion workflow. However, it will not hurt to add this conditional check - ignore if the lock has a machine status of "completed".

Reply:
Sorry, I think I mixed 2 things in (3).
3a) completed locks will still mark the lock as being held, when we should ignore them
3b) 2 units completing at the same time will see the other one as incomplete, and thus not unpin, but both move to completed, and then nothing will unpin.